_posts/2020-02-03-apache-beam-how-beam-runs-on-top-of-flink.md

```diff
@@ -0,0 +1,157 @@
+
+- Docker-based (the default)
+- Process-based (a simple process is started)
+- Externally-provided (K8s or other schedulers)
+- Embedded (intended for testing and only works with Java)
+
+Environments hold the _SDK Harness_ which is the code that handles the execution and the communication with the Runner over the Fn API. For example, when Flink executes Python code, it sends the data to the Python environment containing the Python SDK Harness. Sending data to an external process involves a minor overhead which we have measured to be 5-10% slower than the classic Java pipelines. However, Beam uses fusion of transforms to execute as many transforms as possible in the same environment which share the same input or output. That's why, In real-world scenarios the overhead could be much lower.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/flink-runner-beam-language-portability-architecture.png" width="600px" alt="Language Portability Architecture in beam"/>
+</center>
+
+
+Note: 
+- TODO Mention TFX
+- TODO cross-language
+
+
+## Conclusion
+
+Beam with Flink combines the power of Flink with the flexibility of Beam. All it takes to run Beam is a Flink cluster, which you may already have. Apache Beam's fully-fledged Python API is probably the most compelling argument for using Beam with Flink, but the unified API which allows to "write-once" and "execute-anywhere" is also very appealing to Beam users. On top of this, features like side inputs and a rich connector ecosystem are also reasons why people like Beam. With the introduction of schemas, Beam is heading in a similar direction as Flink with its table API. Speaking of, the next Flink release will include a Python version of the Table API which is based on the language portability of Beam. Looking ahead, the Beam community plans to extend the support for interactive programs like notebooks. TFX is a very powerful way to solve many problems around training and validating machine learning models. For many years, Beam and Flink have inspired and learned from each other. With the Python support being based on Beam in Flink, they only seem to come closer to each other. That's all the better for the community because users have more options and functionality to choose from.
```

Comment:
```suggestion
Using Apache Beam with Apache Flink combines (a.) the power of Flink with (b.) the flexibility of Beam. All it takes to run Beam is a Flink cluster, which you may already have. Apache Beam's fully-fledged Python API is probably the most compelling argument for using Beam with Flink, but the unified API which allows to "write-once" and "execute-anywhere" is also very appealing to Beam users. On top of this, features like side inputs and a rich connector ecosystem are also reasons why people like Beam. With the introduction of schemas, Beam is heading in a similar direction as Flink with its Table API. The latest Flink 1.10 release introduced a Python version of the Table API which is based on the language portability of Beam. Looking ahead, the Beam community plans to extend the support for interactive programs like notebooks. TFX is a very powerful way to solve many problems around training and validating machine learning models. For many years, Beam and Flink have inspired and learned from each other. With the Python support being based on Beam in Flink, they only seem to come closer to each other. That's all the better for the community because users have more options and functionality to choose from.
```

Reply:
This is an outdated diff. I cannot incorporate this suggestion from Github. Please comment only on the current version.

Reply:
sorry @mxm 
probably I added the comments to an old diff. 
I will transfer them to the current one now. 

Thanks 
Markos